### PR TITLE
Allow purging toolshed repository white ghosts using API.

### PR DESCRIPTION
**Pull request by blankenberg**

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py [hunk 1]
```diff
@@ -704,6 +704,45 @@ def repair_repository_revision(self, trans, payload, **kwd):
         # Display the list of repaired repositories.
         return tool_shed_repositories
 
+    @expose_api
+    @web.require_admin
+    def purge_repository(self, trans, id=None, **kwd):
+        """
+        Purge a "white ghost" repository from the database.
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py:711]: I know that's what we have in the equivalent controller method, but "white ghost" is not great as description.
```suggestion
        Purge a repository that is stuck in "NEW" state from the database.
```

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py [hunk 2]
```diff
@@ -704,6 +704,45 @@ def repair_repository_revision(self, trans, payload, **kwd):
...
+
+        GET /api/tool_shed_repositories/{id}/purge
+
+        :param id:  encoded repository id. Either id or name, owner, changeset_revision and tool_shed_url need to be supplied
+        :param kwd: 'name'   : Repository name
+                    'owner'  : Repository owner
+                    'changeset_revision': Changeset revision to uninstall
```
Comment (dannon) [inline lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py:718]:
```suggestion
                    'changeset_revision': Changeset revision to purge
```

File: lib/galaxy/webapps/galaxy/buildapp.py [hunk 3]
```diff
@@ -960,6 +960,12 @@ def populate_api_routes(webapp, app):
                           action='status',
                           conditions=dict(method=["GET"]))
 
+    webapp.mapper.connect('tool_shed_repository',
+                          '/api/tool_shed_repositories/:id/purge',
+                          controller='tool_shed_repositories',
+                          action='purge_repository',
+                          conditions=dict(method=["GET", "POST"]))
```
Comment (dannon) [inline lib/galaxy/webapps/galaxy/buildapp.py:967]: Probably shouldn't allow purges via GET -- can we swap this to POST only?

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py (hunk 1 above)
Comment (jmchilton) [reply lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py:711]: +1

**Comment by nsoranzo** [comment]
Can't we just modify the `ToolShedRepositoriesController.uninstall_repository()` method to call `irm.purge_repository()` instead of `irm.uninstall_repository()` if `repository.is_new` is True?

That would mean using the correct REST method, i.e. `DELETE`.

**Comment by blankenberg** [comment]
I think these are all great points, and making DELETE work as indicated sounds like a smart direction to move in. I likely wont have time to do this in the near future however, so please feel free to close this PR. 

(Also, I've already used these changes to delete the ~200 or so items that were 'annoying' me on one particular instance, so there is no real rush from my end)

It might be useful to sketch out this API in general, before any additional implementation.

**Comment by dannon** [comment]
I really like @nsoranzo's suggestion, too.  I guess let's close this one out as you suggest @blankenberg and revisit that approach.

**Pull request closed without merging**